apps/studio/src/server/modules/searchsg/searchsg.service.ts

```diff
@@ -0,0 +1,125 @@
+import wretch from "wretch"
+
+import { env } from "~/env.mjs"
+
+const SEARCHSG_BASE_URL = "https://api.services.search.gov.sg/admin"
+
+const ISOMER_UA = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) isomer"
+const SearchSgApi = {
+  Auth: `/v1/auth/token`,
+  App: `/v1/bootstrap/applications`,
+} as const
+
+const generateSearchSGParams = ({
+  name,
+  config,
+  url,
+}: UpdateSearchSGConfigProps & {
+  config: SearchSGConfig["data"]
+  url: string
+}) => {
+  return {
+    name: encodeURIComponent(name),
+    tenant: {
+      adminList: ["isomer@open.gov.sg"],
+    },
+    index: {
+      dataSource: {
+        web: [
+          {
+            domain: `https://${url}`,
+            documentIndexConfig: {
+              indexWhitelist: [],
+              indexBlacklist: [],
+            },
+          },
+        ],
+        api: [],
+      },
+    },
+    application: {
+      siteDomain: url,
+      environment: "production",
+      config: {
+        search: {
+          theme: {
+            primary: config.application.config.search.theme.primary,
+            fontFamily: "Inter",
+          },
+        },
+      },
+    },
+  }
+}
+
+interface UpdateSearchSGConfigProps {
+  name: string
+}
+
+const requestSearchSGClient = async () => {
+  const { accessToken, tokenType } = await wretch(
+    `${SEARCHSG_BASE_URL}${SearchSgApi.Auth}`,
+  )
+    .auth(`Basic ${env.SEARCHSG_API_KEY}`)
+    .headers({
+      "User-Agent": ISOMER_UA,
+    })
+    .post()
+    .json<{ accessToken: string; tokenType: string }>()
+
+  return wretch(`${SEARCHSG_BASE_URL}${SearchSgApi.App}`)
+    .auth(`${tokenType} ${accessToken}`)
+    .headers({
+      "Content-Type": "application/json",
+      "User-Agent": ISOMER_UA,
+    })
+}
+
+export const updateSearchSGConfig = async (
+  props: UpdateSearchSGConfigProps,
+  searchsgClientId: string,
+  url: URL,
+) => {
+  const client = await requestSearchSGClient()
+
+  // NOTE: doing fetch before post to avoid cases
+  // where the search domain and data domains
+  // are not direct deriviatives of `site.url`
+  const { data: config } = await client
+    .get(`/${searchsgClientId}`)
+    .json<SearchSGConfig>()
+
+  const updatedConfig = generateSearchSGParams({
+    ...props,
+    config,
+    url: url.host,
```

Comment:
nit: should we pass in the `url` as-is and do the `.host` within `generateSearchSGParams`? a bit weird that we append the `https://` back inside that also haha

Reply:
oh that's a good point, lemme update